lib/text/ui_text_displayer.js

```diff
@@ -189,6 +189,133 @@ shaka.text.UITextDisplayer = class {
+        i++;
+        let wellTerminated = false;
+        for (;i < str.length; i++) {
+          currentValue += str[i];
+          if (str[i] === '>') {
+            wellTerminated = true;
+            break;
+          }
+        }
+
+        if (!wellTerminated || currentValue.length < 3) {
+          // No termination found, append literal string.
+          tokens.push({type: 'text', value: currentValue});
+        } else {
+          let tagType = 'startTag';
+          currentValue = currentValue.substr(1, currentValue.length - 2);
+          if (currentValue[0] == '/') {
+            currentValue = currentValue.substr(1); // Removes /
+            tagType = 'endTag';
+          }
+          if (currentValue) {
+            tokens.push({type: tagType, value: currentValue});
+          }
+        }
+        currentValue = '';
+      } else {
+        currentValue += str[i];
+      }
+    }
+
+    // Final flush.
+    if (currentValue) {
+      tokens.push({type: 'text', value: currentValue});
+    }
+
+    return tokens;
+  }
+
+  /**
+   * Returns a safe collection of html elements for display as a single cue.
+   *
+   * @param {!Array<{type: string, value: string}>} tokens
+   * @return {!Array<HTMLElement>}
+   * @private
+   */
+  buildCueElements_(tokens) {
+    const supportedTags = [
+      'i',
+      'b',
+      'u',
+    ];
+
+    const elements = [];
+
+    // Build a set of root elements from the token collection.
+    const stack = [];
+    for (const currentToken of tokens) {
+      if (currentToken.type === 'text') {
+        // Element is a text field.
+        const textElement = document.createElement('span');
+        textElement.textContent = currentToken.value;
+        if (stack.length > 0) {
+          // Add to bottom stack element.
+          stack[stack.length - 1].appendChild(textElement);
+        } else {
+          // Add as standalone.
+          elements.push(textElement);
+        }
+      } else if (supportedTags.includes(currentToken.value)) {
+        // Element is a supported tag.
+        if (currentToken.type === 'startTag') {
+          // Add this new element to the stack.
+          stack.push(document.createElement(currentToken.value));
+          if (stack.length > 1) {
+            // Add this item as a child of the element one up the stack.
+            stack[stack.length - 2].appendChild(stack[stack.length - 1]);
+          }
+        } else if (currentToken.type === 'endTag') {
+          // Check to see if this closes the last opened tag.
+          const bottomName = stack[stack.length - 1].tagName.toLowerCase();
+          if (bottomName === currentToken.value.toLowerCase()) {
+            // Close this tag off the stack.
+            const v = stack.pop();
+            if (stack.length === 0) {
+              // Stack is empty, push root of this tree into elements.
+              elements.push(v);
+            }
+          } else {
+            // Input syntax error.
```

Comment:
It seems like this should never happen (i.e. an endTag closing something that other than the last opened tag), and it jeopardizes the validity of the entire markup line. Perhaps it would be worth throwing an error here, or returning a blank array to show that the line couldn't be built.

Reply:
I changed it!